salt/client/__init__.py

```diff
@@ -1986,13 +1993,98 @@ def pub_async(
+            try:
+                # Ensure that the event subscriber is connected.
+                # If not, we won't get a response, so error out
+                if listen and not self.event.connect_pub(timeout=timeout):
+                    raise SaltReqTimeoutError()
+                log.warn("SENDING %r", payload_kwargs)
+                payload = yield channel.send(payload_kwargs, timeout=timeout)
+            except SaltReqTimeoutError:
+                raise SaltReqTimeoutError(
+                    "Salt request timed out. The master is not responding. You "
+                    "may need to run your command with `--async` in order to "
+                    "bypass the congested event bus. With `--async`, the CLI tool "
+                    "will print the job id (jid) and exit immediately without "
+                    "listening for responses. You can then use "
+                    "`salt-run jobs.lookup_jid` to look up the results of the job "
+                    "in the job cache later."
+                )
+
+            if not payload:
+                # The master key could have changed out from under us! Regen
+                # and try again if the key has changed
+                key = self.__read_master_key()
+                if key == self.key:
+                    raise salt.ext.tornado.gen.Return(payload)
+                self.key = key
+                payload_kwargs["key"] = self.key
+                payload = yield channel.send(payload_kwargs)
+
+            error = payload.pop("error", None)
+            if error is not None:
+                if isinstance(error, dict):
+                    err_name = error.get("name", "")
+                    err_msg = error.get("message", "")
+                    if err_name == "AuthenticationError":
+                        raise AuthenticationError(err_msg)
+                    elif err_name == "AuthorizationError":
+                        raise AuthorizationError(err_msg)
+
+                raise PublishError(error)
+
+            if not payload:
+                raise salt.ext.tornado.gen.Return(payload)
+
+        raise salt.ext.tornado.gen.Return(
+            {"jid": payload["load"]["jid"], "minions": payload["load"]["minions"]}
+        )
+
+    @salt.ext.tornado.gen.coroutine
+    def arbitrary_pub_async(
+        self,
+        tgt,
+        payload,
+        tgt_type="glob",
+        timeout=5,
+        io_loop=None,
+        listen=True,
+        cmd=None,
+        **kwargs
+    ):
+        """
+        Publish an arbitrary payload. Maybe this is what events do?
+        """
+        if self.opts.get("ipc_mode", "") != "tcp" and not os.path.exists(
+            os.path.join(self.opts["sock_dir"], "publish_pull.ipc")
+        ):
+            log.error(
+                "Unable to connect to the salt master publisher at %s",
+                self.opts["sock_dir"],
+            )
+            raise SaltClientError
+        payload_kwargs, payload = payload, None
```

Comment:
I'm probably going to reorder these if it turns out that I have to have them.

I *think* I do, because `pub_async` doesn't actually offer the ability to publish more or less arbitrary data. And maybe this is the wrong way to communicate! So if anyone has some better ideas, I'm all ears.